Data.Rmd

```diff
@@ -0,0 +1,567 @@
+
+```{r, eval = FALSE}
+mfl()
+#> [1] "a" "b" "c"
+
+set_mfl(c("j", "f", "b"))
+#> Error in set_mfl() : 
+#>   cannot change value of locked binding for 'favorite_letters'
+```
+
+Because `favorite_letters` is a regular character vector, modification requires making a copy and rebinding the name `favorite_letters` to this new value.
+And that is what's disallowed: you can't change the binding for objects in the package namespace (well, at least not without trying harder than this).
+Defining `favorite_letters` this way only works if you will never need to modify it.
+
+However, if we maintain state within an internal package environment, we **can** modify objects contained in the environment (and even add completely new objects).
+Here's an alternative implementation that uses an internal environment named "the".
+
+```{r, eval = FALSE}
+the <- new.env(parent = emptyenv())
+the$favorite_letters <- letters[1:3]
+
+#' Report my favorite letters
+#' @export
+mfl2 <- function() {
+  the$favorite_letters
+}
+
+#' Change my favorite letters
+#' @export
+set_mfl2 <- function(l = letters[24:26]) {
+  old <- the$favorite_letters
+  the$favorite_letters <- l
+  invisible(old)
+}
+```
+
+Now a user *can* register their favorite letters:
+
+```{r, eval = FALSE}
+mfl2()
+#> [1] "a" "b" "c"
+
+set_mfl2(c("j", "f", "b"))
+#> [1] "j" "f" "b"
+
+mfl2()
+#> [1] "j" "f" "b"
+```
+
+Note that this new value for `the$favorite_letters` persists only for the remainder of the current R session (or until the user calls `set_mfl2()` again).
+More precisely, the altered state persists only until the next time the package is loaded (including via `load_all()`).
+At load time, the environment `the` is reset to an environment containing exactly one object, named `favorite_letters`, with value ("a", "b", "c").
+It's like the movie Groundhog Day.
+(We'll discuss more persistent package- and user-specific data in the next section.)
+
+Jim Hester introduced our group to the nifty idea of using "the" as the name of an internal package environment.
+This lets you to refer to the objects inside in a very natural way, such as `the$token`, meaning "the token".
+It is also important to specify `parent = emptyenv()` when defining an internal environment, as you generally don't want the environment to inherit from any other (nonempty) environment.
+
+As seen in the example above, the definition of the environment should happen as a top-level assignment in a file below `R/`.
+(In particular, this is a legitimate reason to define a non-function at the top-level of a package; see section @sec-code-when-executed for why this should be rare.) As for where to place this definition, there are two considerations:
+
+-   Define it before you use it.
+    If other top-level calls refer to the environment, the definition must come first when the package code is being executed at build time.
+    This is why `R/aaa.R` is a common and safe choice.
```

Comment:
To get around the need to put it in `R/aaa.R`, in dplyr we use a trick where `the <- new_environment()` can be put anywhere, but when we have top-level calls that refer to it, we wrap them in `rlang::on_load()` to run those expressions in the `.onLoad()` hook (which is after `the` has been created)

I think this idea is experimental enough that your current recommendation is best, but thought I'd mention it

https://cs.github.com/tidyverse/dplyr/blob/5aa0b7c4fcd91f87ce0f76821d5fe6c137b4e8a0/R/dplyr.R#L23

https://cs.github.com/tidyverse/dplyr/blob/5aa0b7c4fcd91f87ce0f76821d5fe6c137b4e8a0/R/conditions.R#L287

Reply:
Yeah I feel like we have quite of bit of interesting "on load" stuff right now. But that it's also a bit too experimental (and getting into the weeds) to work into the book today.